### PR TITLE
[OCDM] Add functionality to construct Session private data.

### PR DESCRIPTION
**Pull request by sramani-metro**
Session private data like SVP hardware specific context can be constructed and associated with a OCDM Session.

### CODE REVIEW

File: Source/ocdm/adapter/gstreamer/open_cdm_adapter.cpp [hunk 1]
```diff
@@ -39,6 +39,19 @@ inline bool mappedBuffer(GstBuffer *buffer, bool writable, uint8_t **data, uint3
     return true;
 }
 
+uint32_t opencdm_construct_session_private(struct OpenCDMSession* session, void* &pvtData)
```
Comment (pwielders) [inline Source/ocdm/adapter/gstreamer/open_cdm_adapter.cpp:42]: I guess this is not needed as the static will look for these methods and if the methods are not there just set the pointer to null ???? Right ?

File: Source/ocdm/open_cdm_impl.cpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                Load();
+        }
+        ~SessionPrivate() = default;
+
+        OpenCDMError Construct(struct OpenCDMSession* session, void* &pvtData)
+        {
+            OpenCDMError result = OpenCDMError::ERROR_METHOD_NOT_IMPLEMENTED;
+            if(_constructSessionPvt != nullptr) {
+                result = ( _constructSessionPvt(session, pvtData) == 0 ? OpenCDMError::ERROR_NONE : OpenCDMError::ERROR_UNKNOWN);
+            }
```
Comment (pwielders) [inline Source/ocdm/open_cdm_impl.cpp:44]: if there would be an else here where the pbtData is set to null, one could live without having the:
EXTERNAL uint32_t opencdm_construct_session_private(struct OpenCDMSession* session, void* &pvtData);
or 
EXTERNAL uint32_t opencdm_destruct_session_private(struct OpenCDMSession* session, void* &pvtData); 
methods in the adapter. So only for the Realtek/Amlogic adapters you need to add those, all other variant can stay as they where.
If we would implement them we could just link against them at compile time as we would expect them than. It ould make the code faster but would be not backwards compatible.... It is what you prefer... Lets discuss on skype later ...

File: Source/ocdm/open_cdm_impl.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        *session = new OpenCDMSession(system, std::string(initDataType),
+                            initData, initDataLength, CDMData,
+                            CDMDataLength, licenseType, callbacks, userData);
+        result = (*session != nullptr ? OpenCDMError::ERROR_NONE
+                                      : OpenCDMError::ERROR_INVALID_SESSION);
+        if(result == OpenCDMError::ERROR_NONE) {
+            void *pvtData = nullptr;
+            (*session)->_pvtData = nullptr;
+            OpenCDMError pvt_result(OpenCDMError::ERROR_METHOD_NOT_IMPLEMENTED);
+            pvt_result = SessionPvt.Construct(*session, pvtData);
```
Comment (pwielders) [inline Source/ocdm/open_cdm_impl.cpp:94]: Is there a reason for not directly pushing the (*session)->_pvtData ? You could just call: 
pvt_result = SessionPvt.Construct(*session,  (*session)->_pvtData); 
Than you could also drop the following code: 
`
if(pvt_result == OpenCDMError::ERROR_NONE) {
    (*session)->_pvtData = pvtData;
}
else 
`

**Review by pwielders — COMMENTED** [review]
Looks good and indeed decouples/moves the platform specific code back into the adapter.
Some minor remarks:
1) I guess we do not need a default implementation (it looks for an implementation)
2) Question/small optimization on calling the set data.

2 is not really important I think on 1 we should decode to go left or right, Always implement and hook it up link time or only implement if needed and not link time resolve but runtime.

File: Source/ocdm/adapter/gstreamer/open_cdm_adapter.cpp (hunk 1 above)
Comment (sramani-metro) [reply Source/ocdm/adapter/gstreamer/open_cdm_adapter.cpp]: Yes, I agree. Removed

File: Source/ocdm/open_cdm_impl.cpp (hunk 2 above)
Comment (sramani-metro) [reply Source/ocdm/open_cdm_impl.cpp:44]: Updated.

File: Source/ocdm/open_cdm_impl.cpp (hunk 3 above)
Comment (sramani-metro) [reply Source/ocdm/open_cdm_impl.cpp]: Updated based on previous comment.

**Review by pwielders — DISMISSED** [review]

**Review by pwielders — APPROVED** [review]

**Pull request merged**